code/modules/antagonists/heretic/heretic_antag.dm

```diff
@@ -0,0 +1,682 @@
+		knowledge_data["gainFlavor"] = found_knowledge.gain_text
+		knowledge_data["cost"] = found_knowledge.cost
+		knowledge_data["hereticPath"] = found_knowledge.route
+		knowledge_data["color"] = path_to_color[found_knowledge.route] || "grey"
+
+		data["learnedKnowledge"] += list(knowledge_data)
+
+	return data
+
+/datum/antagonist/heretic/ui_static_data(mob/user)
+	var/list/data = list()
+
+	data["total_sacrifices"] = total_sacrifices
+	data["ascended"] = ascended
+	data["objectives"] = get_objectives()
+
+	return data
+
+/datum/antagonist/heretic/ui_act(action, params)
+	. = ..()
+	if(.)
+		return
+
+	switch(action)
+		if("research")
+			var/datum/heretic_knowledge/researched_path = text2path(params["path"])
+			if(!ispath(researched_path))
+				CRASH("Heretic attempted to learn non-heretic_knowledge path! (Got: [researched_path])")
+
+			if(initial(researched_path.cost) > knowledge_points)
+				return
+			if(!gain_knowledge(researched_path))
+				return
+
+			knowledge_points -= initial(researched_path.cost)
+			return TRUE
+
+/datum/antagonist/heretic/ui_status(mob/user, datum/ui_state/state)
+	if(user.stat == DEAD)
+		return UI_CLOSE
+	return ..()
+
+/datum/antagonist/heretic/greet()
+	owner.current.playsound_local(get_turf(owner.current), 'sound/ambience/antag/ecult_op.ogg', 100, FALSE, pressure_affected = FALSE, use_reverb = FALSE)//subject to change
+	to_chat(owner, "<span class='boldannounce'>You are the Heretic!</span><br>\
+	<B>The old ones gave you these tasks to fulfill:</B>")
+	owner.announce_objectives()
+	to_chat(owner, "<span class='cult'>The book whispers, the forbidden knowledge walks once again!<br>\
+	The Forbidden Knowledge panel allows you to research abilities, read it very carefully! You cannot undo what has been done!<br>\
+	You gain charges by either collecting influences or sacrificing people tracked by the living heart<br> \
+	You can find a basic guide at: https://wiki.beestation13.com/view/Heretics </span>")
+	owner.current.client?.tgui_panel?.give_antagonist_popup("Heretic",
+		"Collect influences or sacrifice targets to expand your forbidden knowledge.")
+
+/datum/antagonist/heretic/farewell()
+	if(!silent)
+		to_chat(owner.current, "<span class='userdanger'>Your mind begins to flare as the otherworldly knowledge escapes your grasp!</span>")
+	return ..()
+
+/datum/antagonist/heretic/on_gain()
+	if(isipc(owner.current))//Due to IPCs having a mechanical heart it messes with the living heart, so no IPC heretics for now
+		var/mob/living/carbon/C = owner.current	//only carbons have dna now, so we have to typecaste
+		C.set_species(/datum/species/human)
+		var/replacementName = random_unique_name(C.gender)
+		if(C.client.prefs.active_character.custom_names["human"])
+			C.fully_replace_character_name(C.real_name, C.client.prefs.active_character.custom_names["human"])
+		else
+			C.fully_replace_character_name(C.real_name, replacementName)
+	if(give_objectives)
+		forge_primary_objectives()
```

Comment:
Are there non primary objectives? Doesn't look like it, this should be named the same as in other antagonists

Reply:
Renamed